### PR TITLE
Indexer Messaging and Error Improvements

### LINKED ISSUE #5658: Capture Known Indexer HTTP Errors and Provide Better Messaging to users
### Is there an existing issue for this?

- [X] I have searched the existing open and closed issues

### Is your feature request related to a problem? Please describe

Indexer servers may be down and generic "unable to connect to indexer xxxxxx" messaging is not clear to users that the server is unavailable.

### Describe the solution you'd like

- Add IsServerNotAvailable bool to HTTPResponse for common server is down codes like 502, 504, 503, 522, 524
- Improve ValidationFailure and Logging messaging to indicate that the server is not available

### Describe alternatives you've considered

Continue to provide users with technical errors that users don't understand

### Anything else?

Similar to https://github.com/Prowlarr/Prowlarr/pull/1666/files

Expect this to be closed per https://github.com/Prowlarr/Prowlarr/pull/1666#discussion_r1197831516 as it appears there was no upstream appetite for this.

### PR DESCRIPTION
**Pull request by bakerboy448**
#### Database Migration
NO

#### Description
- New: Improve Indexer Error Handling and Messaging
- Fixed: (Newznab Torznab) Better HTTP Error Handling

Catch the following expected HTTP responses and pass messaging indicating the indexer is down and try again later as bool IsServerNotAvailable for HTTP Response

HTTP 500 +

#### Todos
- [ ] Tests
- [ ] Wiki Updates

#### Issues Fixed or Closed by this PR

* Closes #5658
* Ref https://github.com/Prowlarr/Prowlarr/pull/1672

### CODE REVIEW

File: src/NzbDrone.Core/Indexers/RssParser.cs [hunk 1]
```diff
@@ -120,7 +120,8 @@ protected virtual ReleaseInfo CreateNewReleaseInfo()
 
         protected virtual bool PreProcess(IndexerResponse indexerResponse)
         {
-            if (indexerResponse.HttpResponse.StatusCode != HttpStatusCode.OK)
+            // Server Down HTTP Errors are handled in HTTPIndexerBase so ignore them here
+            if (indexerResponse.HttpResponse.StatusCode != HttpStatusCode.OK && !indexerResponse.HttpResponse.IsServerNotAvailable)
```
Comment (markus101) [inline src/NzbDrone.Core/Indexers/RssParser.cs:124]: Don't we still want to throw if we got an unexpected response instead of trying to process responses?

**Review by markus101 — CHANGES_REQUESTED** [review]

File: src/NzbDrone.Core/Indexers/RssParser.cs (hunk 1 above)
Comment (bakerboy448) [reply src/NzbDrone.Core/Indexers/RssParser.cs:124]: they were getting thrown in httpindexerbase when testing in downstream; should be same here

**Comment by bakerboy448** [comment]
can/will squash once review and feedback

File: src/NzbDrone.Core/Indexers/RssParser.cs (hunk 1 above)
Comment (markus101) [reply src/NzbDrone.Core/Indexers/RssParser.cs:124]: Right, but if they're getting thrown and still reaching here then we shouldn't try to process the response.

Comment (bakerboy448) [reply src/NzbDrone.Core/Indexers/RssParser.cs:124]: @mynameisbogdan any insight here

quick testing has this concept working fine in prowlarr but doesn't really work in sonarr (assume same for downstream starr)

File: src/NzbDrone.Common/Http/HttpResponse.cs [hunk 2]
```diff
@@ -50,6 +50,8 @@ public string Content
 
         public bool HasHttpError => (int)StatusCode >= 400;
 
+        public bool IsServerNotAvailable => (int)StatusCode >= 500;
```
[the remark this answers is not in the document]
Comment (markus101) [reply src/NzbDrone.Common/Http/HttpResponse.cs:53]: That does fit better with the both the existing `HasHttpError` and the fact that a 500 error doesn't mean the server isn't available, it means it failed to handle the request.

[the remark this answers is not in the document]
Comment (bakerboy448) [reply src/NzbDrone.Common/Http/HttpResponse.cs]: push was coming :)

File: src/NzbDrone.Core/Indexers/RssParser.cs (hunk 1 above)
Comment (bakerboy448) [reply src/NzbDrone.Core/Indexers/RssParser.cs]: little lost in the weeds for where exactly to apply that

**Review by markus101 — APPROVED** [review]

File: src/NzbDrone.Core/Indexers/HttpIndexerBase.cs [hunk 3]
```diff
@@ -397,6 +400,53 @@ protected virtual ValidationFailure TestConnection()
...
+                if (ex.Response.HasHttpServerError)
+                {
+                    return new ValidationFailure(string.Empty, "Unable to connect to indexer, indexer's server is unavailable. Try again later. " + ex.Message);
+                }
+
+                if (ex.Response.StatusCode is HttpStatusCode.Forbidden or HttpStatusCode.Unauthorized)
+                {
+                    return new ValidationFailure(string.Empty, "Unable to connect to indexer, invalid credentials. " + ex.Message);
+                }
+                return new ValidationFailure(string.Empty, "Unable to connect to indexer, check the log above the ValidationFailure for more details. " + ex.Message);
```
Comment (markus101) [inline src/NzbDrone.Core/Indexers/HttpIndexerBase.cs:422]: Needs new line before return

**Review by markus101 — APPROVED** [review]
CI is failing, but changes look good once CI is happy.

**Pull request merged**